packages/@aws-cdk/aws-cognito-identitypool/lib/identitypool-user-pool-authentication-provider.ts

```diff
@@ -0,0 +1,85 @@
+import { randomBytes } from 'crypto';
+import {
+  IUserPoolIdentityProvider,
+  UserPoolClient,
+  UserPoolClientProps,
+} from '@aws-cdk/aws-cognito';
+import { DependableTrait } from '@aws-cdk/core';
+
+/**
+ * Represents a UserPoolAuthenticationProvider
+ */
+export interface IUserPoolAuthenticationProvider {
+  /**
+   * Client Id of the Associated User Pool Client
+   */
+  readonly clientId: string
+
+  /**
+   * Whether to disable the pool's default server side token check
+   */
+  readonly disableServerSideTokenCheck: boolean;
+
+  /**
+   * The identity providers associated with the UserPool
+   */
+  readonly identityProviders: IUserPoolIdentityProvider[];
+}
+
+/**
+ * Props for the User Pool Authentication Provider
+ */
+export interface UserPoolAuthenticationProviderProps extends UserPoolClientProps {
+
+  /**
+   * Setting this to true turns off identity pool checks for this user pool to make sure the user has not been globally signed out or deleted before the identity pool provides an OIDC token or AWS credentials for the user
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cognito-identitypool-cognitoidentityprovider.html
+   * @default false
+   */
+  readonly disableServerSideTokenCheck?: boolean;
+}
+
+/**
+ * Defines a User Pool Authentication Provider
+ */
+export class UserPoolAuthenticationProvider implements IUserPoolAuthenticationProvider {
+
+  /**
+   * Client Id of the Associated User Pool Client
+   */
+  public readonly clientId: string
+
+  /**
+   * Whether to disable the pool's default server side token check
+   */
+  public readonly disableServerSideTokenCheck: boolean;
+
+  /**
+   * The identity providers associated with the UserPool
+   */
+  public readonly identityProviders: IUserPoolIdentityProvider[] = [];
+
+  constructor(props: UserPoolAuthenticationProviderProps) {
+    const client = this.configureUserPoolClient(props);
+    this.clientId = client.userPoolClientId;
+    this.identityProviders = props.userPool.identityProviders;
+    this.disableServerSideTokenCheck = props.disableServerSideTokenCheck ? true : false;
+    DependableTrait.implement(this, {
```

Comment:
@smguggen I'm not sure I understand what this is doing/why this is needed?

Reply:
It's configuring a user pool/user pool client so that it can be accepted by the `CFNIdentityPool`'s `cognitoIdentityProviders` field. [Here's](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cognito-identitypool-cognitoidentityprovider.html) the documentation for it.

Reply:
And the `DependableTrait` part is so it can be set as a dependency of the identity pool. I added that after I ran into a timing issue with an identity pool looking for a user pool that hadn't been created yet.

Reply:
@smguggen sorry, I was asking specifically about the `DependableTrait.implement()`

Reply:
@corymhall So the class configures a user pool and a user pool client for the identity pool and can be generated directly by the class, but if you're trying to put a dependency on the user pool or the user pool client, you can't because it's hidden inside the class. So `implement` just says that when a dependency is placed on the class, that means to pass that dependency on to both the user pool and the user pool client inside of it.

Reply:
`DependableTrait` should only be used for `Constructs`, really. You're not supposed to be able to go `addDependency(randomClass)` (even though you could).

If you need to have side effects happen when this class is used in an `IdentityPool`, use the `bind()` pattern we use for that. Something like:

```ts
public bind(scope: Construct, identityPool: IdentityPool) {
  Node.of(identityPool).addDependency(this.client);
}
```

Reply:
@rix0rrr Thanks for this, looking at some other examples of this it seems like this was the missing piece I was looking for in this class. Judging from the way this is used elsewhere, I should probably return the configured properties in the bind method and then just require the bind method in `IUserPoolAuthenticationProvider`. Implementing now, will alert after tests finish.

Reply:
@rix0rrr bind method implemented and tests adjusted